### PR TITLE
Fix performance, improve sensor update and calculation

### PR DESCRIPTION
**Pull request by isorin**
------------------------ 2nd commit ----------------

Performance enhancements 2nd stage & refactor sensor calculation.
- limit forecast list scan for each sensor
- split sensors in two groups (based on type of update required)
- calculate some sensors more precisely

------------------------ 1st commit ----------------

This is a 1st attempt (rough stage) to fix some performance issues with the code.
I intend to follow-up with a couple of more optimisations.

I also fixed a bug with code that attempted to remove elements from a list while iterating through it. This results in skipping elements as the entire list shifts. (I didn't check to see if there are other places in the code where this technique is used)

The code is highly inefficient.
Basically, in my instance, it iterates over 4000 past records (about 3 months worth of forecasts) every 1 minute for multiple sensors as "self._data_forecasts" contains all past records (at some stage it will hold 2 years worth of data - do we need 2 years worth of solar forecasts, we could allow users to configure a history limit).

From what I see it doesn't need past forecast records for any of the sensors so I am proposing the following changes to make it more efficient (by my calculations it should reduce complexity by a factor of 50x):

1. Remove past forecasts from self._data_forecasts
2. Reduce update frequency for all sensors to 5 minutes (do we need solar forecast updates every minute?)

Future performance improvement attempts should exploit the following facts:
- self._data_forecasts is ordered so some sensors can only loop until "period_start" becomes > X
- some sensors don't need recalculation (for instance, future days) or don't need to be recalculated every 5 minutes

Also, I looked a bit at the way calculations are made for some sensors, for instance power_next_30mins.
There is no "intelligence" in the way the sensor is calculated. It should take the remaining minutes (let's say A) from the current 30 minutes interval and the rest of the minutes (30-A) from the next 30 minutes interval and use a formula such as:
power_next_30mins = (A / 30) * power_current_30mins_interval + (1 - A / 30) * power_next_30mins_interval
to try and estimate the power available in the next 30 minutes.

Anyway, all in good time.

Here are my proposed changes, for your consideration.

### CODE REVIEW

File: custom_components/solcast_solar/solcastapi.py [hunk 1]
```diff
@@ -615,33 +622,35 @@ async def http_data_call(self, r_id = None, api = None, dopast = False):
...
+
         for x in _data:
             #loop each rooftop site and its forecasts
             
-            itm = next((item for item in _forecasts if item["period_start"] == x["period_start"]), None)
+            itm = _fcasts_dict.get(x["period_start"])
             if itm:
                 itm["pv_estimate"] = x["pv_estimate"]
                 itm["pv_estimate10"] = x["pv_estimate10"]
                 itm["pv_estimate90"] = x["pv_estimate90"]
```
Comment (tmenguy) [inline custom_components/solcast_solar/solcastapi.py:642]: do you know why here we replace the values in itm ? instead of adding them?

Comment (oziee) [reply custom_components/solcast_solar/solcastapi.py:642]: Why are you asking the same question in multiple places??

Stick to one please or I will block you for spamming

Comment (tmenguy) [reply custom_components/solcast_solar/solcastapi.py:642]: I realized opening an issue was a bad idea, should have closed the issue … sorry again.

**Comment by isorin** [comment]
Closed, as not wanted/needed.
For reference, see:
https://github.com/isorin/ha-solcast-solar/tree/perf_fix_1

**Comment by isorin** [comment]
Absolutely not, I know his reply was for your comment. I closed the PR as there wasn't any activity on it for a long time. I'm keeping my fork as I use it and, if @oziee wants/needs any of my changes, he can access and incorporate them directly from my fork.

**Pull request closed without merging**